test/test_torch.py

```diff
@@ -2466,39 +2466,179 @@ def _test_rpow(self, cast):
+            exps = [0, 1, 2, 4,
+                    torch.tensor((3, 3), dtype=dt, device=dev).random_(0, 5)]
+            for exp in exps:
+                _check_against_np(tensor, exp)
 
+        _test_integral_pow(torch.int8, (-3, 4), device)
+        _test_integral_pow(torch.uint8, (0, 4), device)
+        _test_integral_pow(torch.int16, (-5, 5), device)
+        _test_integral_pow(torch.int64, (-10, 10), device)
         if not IS_WINDOWS:
-            typecasts.append(lambda x: x.int())
+            _test_integral_pow(torch.int32, (-10, 10), device)
 
-        shape = (11, 5)
-        tensor = cast(torch.LongTensor(shape).random_(-10, 10))
-        exps = [0, 1, 2, 5, cast(torch.LongTensor(shape).random_(0, 20))]
+    @torchtest.for_all_device_types()
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_int_tensor_pow_neg_ints(self, device):
+        ints = [torch.iinfo(torch.int32).min,
+                -3, -2, -1, 0, 1, 2, 3,
+                torch.iinfo(torch.int32).max]
+        neg_ints = [torch.iinfo(torch.int32).min, -3, -2, -1]
 
-        for typecast in typecasts:
-            for exp in exps:
-                t = typecast(tensor)
-                e = exp if isinstance(exp, int) else typecast(exp)
-                check_against_np(t, e)
+        tensor = torch.tensor(ints, dtype=torch.int32, device=device)
+        nparr = np.array(ints, dtype=np.int32)
+        out = torch.empty_like(tensor, device=device)
 
-    def test_int_pow(self):
-        self._test_int_pow(self, lambda x: x)
+        for pow in neg_ints:
+            self.assertRaisesRegex(
+                ValueError,
+                "Integers to negative integer powers are not allowed.",
+                lambda: np.power(nparr, pow))
+            self.assertRaisesRegex(
+                RuntimeError,
+                "Integers to negative integer powers are not allowed.",
+                lambda: tensor.pow(pow))
+            self.assertRaisesRegex(
+                RuntimeError,
+                "Integers to negative integer powers are not allowed.",
+                lambda: tensor.pow_(pow))
+            self.assertRaisesRegex(
+                RuntimeError,
+                "Integers to negative integer powers are not allowed.",
+                lambda: torch.pow(tensor, pow))
+            self.assertRaisesRegex(
+                RuntimeError,
+                "Integers to negative integer powers are not allowed.",
+                lambda: torch.pow(tensor, pow, out=out))
+
+    @torchtest.for_all_device_types()
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_long_tensor_pow_floats(self, device):
+        ints = [0, 1, 23, 4567]
+        floats = [0.0, 1 / 3, 1 / 2, 1.0, 3 / 2, 2.0]
+
+        tensor = torch.tensor(ints, dtype=torch.int64, device=device)
+        nparr = np.array(ints, dtype=np.int64)
+
+        if device == 'cuda' and not TEST_WITH_ROCM:
```

Comment:
We should really split this test into 2; it doesn't make sense for them to be the same function.

Reply:
addressed in different way, I added this:
```
if device == 'cuda' and not TEST_WITH_ROCM:
                # Current pow CUDA implementation casts exponent
                # to tensor dtype, but numpy does not, that's why:
                # pow CUDA  4 ^ 0.5 = 1
                # numpy pow 4 ^ 0.5 = 2
                # This line must be deleted as soon as
                # pow CUDA implementation is fixed.
                expected = np.power(nparr, int(pow)).astype(np.int64)
            else:
                # cpu CUDA implementation is already fixed and
                # does not cast exponent to tensor dtype,
                # that why it is compatible with numpy.
                expected = np.power(nparr, pow).astype(np.int64)
```